### PR TITLE
Check ftyp validity

### PR DESCRIPTION
**Pull request by lvandeve**
Check if the ftyp box is present, at the correct position, and only
once, and the major brand matches "jxl "

This makes the decoder more stringent about this, which aids in
conformance.

### CODE REVIEW

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/decode.cc [hunk 1]
```diff
@@ -1743,6 +1746,23 @@ JxlDecoderStatus JxlDecoderProcessInput(JxlDecoder* dec) {
                sizeof(dec->box_decoded_type));
       }
 
+      // Box order validity checks
+      // The signature box at box_count == 1 is not checked here since that's
+      // already done at the beginning.
+      dec->box_count++;
+      if (dec->box_count == 2) {
```
Comment (deymo) [inline lib/jxl/decode.cc:1753]: can't there be other, non-jxl-related boxes?

Comment (jonsneyers) [reply lib/jxl/decode.cc:1753]: No, ftyp has to be the first box after the signature box.

**Comment by lvandeve** [comment]
I updated this pull request, to make the check less strict: it now no longer requires the exact same 20 bytes

Even though the specification requires those exact 20 bytes, there may be reasons to do some changes to the ftyp box in the future and still have them compatible with existing decoders

**Review by veluca93 — DISMISSED** [review]

File: lib/jxl/decode.cc [hunk 2]
```diff
@@ -1743,6 +1746,17 @@ JxlDecoderStatus JxlDecoderProcessInput(JxlDecoder* dec) {
                sizeof(dec->box_decoded_type));
       }
 
+      // Box order validity checks
+      // The signature box at box_count == 1 is not checked here since that's
+      // already done at the beginning.
+      dec->box_count++;
+      if (dec->box_count == 2 && memcmp(dec->box_type, "ftyp", 4) != 0) {
+        return JXL_API_ERROR("there second box must be the ftyp box");
```
Comment (jonsneyers) [inline lib/jxl/decode.cc:1754]: "the second box"

**Review by jonsneyers — APPROVED** [review]

File: lib/jxl/decode.cc (hunk 2 above)
Comment (lvandeve) [reply lib/jxl/decode.cc]: fixed :)

**Review by veluca93 — APPROVED** [review]

**Pull request merged**